plugin/kvictoria/kvictoria.go

```diff
@@ -0,0 +1,311 @@
+func NewMetrics(namespace string, opts ...Opt) *Metrics {
+	return &Metrics{
+		cfg: newCfg(namespace, opts...),
+	}
+}
+
+// OnNewClient implements the [kgo.HookNewClient] interface for metrics gathering.
+// This method is meant to be called by the hook system and not by the user.
+func (m *Metrics) OnNewClient(client *kgo.Client) {
+	clientID := client.OptValue(kgo.ClientID).(string)
+
+	if m.clientID != "" {
+		panic(fmt.Errorf("do not reuse a Metrics instance for multiple kgo clients; current client id: %s", m.clientID))
+	}
+
+	m.clientID = clientID
+	m.set = vm.NewSet()
+
+	labels := map[string]string{"client_id": clientID}
+
+	m.set.GetOrCreateGauge(m.buildName("buffered_fetch_bytes_total", labels), func() float64 {
+		return float64(client.BufferedFetchBytes())
+	})
+	m.set.GetOrCreateGauge(m.buildName("buffered_fetch_records_total", labels), func() float64 {
+		return float64(client.BufferedFetchRecords())
+	})
+	m.set.GetOrCreateGauge(m.buildName("buffered_produce_bytes_total", labels), func() float64 {
+		return float64(client.BufferedProduceBytes())
+	})
+	m.set.GetOrCreateGauge(m.buildName("buffered_produce_records_total", labels), func() float64 {
+		return float64(client.BufferedProduceRecords())
+	})
+
+	vm.RegisterSet(m.set)
+}
+
+// OnClientClosed implements the [kgo.HookClientClosed] interface for metrics gathering.
+// This method is meant to be called by the hook system and not by the user.
+//
+// This will unregister all metrics that are scoped to the client id of the client provided.
+func (m *Metrics) OnClientClosed(client *kgo.Client) {
+	vm.UnregisterSet(m.set, true)
+}
+
+// OnBrokerConnect implements the [kgo.HookBrokerConnect] interface for metrics gathering.
+// This method is meant to be called by the hook system and not by the user.
+func (m *Metrics) OnBrokerConnect(meta kgo.BrokerMetadata, dialTime time.Duration, _ net.Conn, err error) {
+	labels := map[string]string{
+		"client_id": m.clientID,
+		"node_id":   kgo.NodeName(meta.NodeID),
+	}
+
+	if err != nil {
+		m.set.GetOrCreateCounter(m.buildName("connect_errors_total", labels)).Inc()
+		return
+	}
+
+	m.set.GetOrCreateCounter(m.buildName("connects_total", labels)).Inc()
+	m.set.GetOrCreateHistogram(m.buildName("connect_seconds", labels)).Update(dialTime.Seconds())
+}
+
+// OnBrokerDisconnect implements the [kgo.HookBrokerDisconnect] interface for metrics gathering.
+// This method is meant to be called by the hook system and not by the user
+func (m *Metrics) OnBrokerDisconnect(meta kgo.BrokerMetadata, _ net.Conn) {
+	labels := map[string]string{
+		"client_id": m.clientID,
+		"node_id":   kgo.NodeName(meta.NodeID),
+	}
+
+	m.set.GetOrCreateCounter(m.buildName("disconnects_total", labels)).Inc()
+}
+
+// OnBrokerWrite is a noop implementation of [kgo.HookBrokerWrite], logic moved to OnBrokerE2E
+func (m *Metrics) OnBrokerWrite(meta kgo.BrokerMetadata, _ int16, bytesWritten int, _, _ time.Duration, err error) {
```

Comment:
You can probably leave the no-op functions out, fwiw. There's no impact to not having them if they're no-ops anyway.